### PR TITLE
Add React.unstable_createEventComponent

### PR DESCRIPTION
**Pull request by necolas**
Introduces a new unstable API for creating event components from event responders.

Ref #15257

### CODE REVIEW

**Comment by necolas** [comment]
The bundle size of `react` and all the `react-event` modules has increased. Not sure what the right way to set this up is

File: packages/react/src/ReactCreateEventComponent.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  props: null | Object,
+  responder: ReactEventResponder,
+};
+
+export function createEventComponent({
+  displayName,
+  props,
+  responder,
+}: EventComponentConfig): ReactEventComponent {
+  const eventComponent = {
```
Comment (trueadm) [inline packages/react/src/ReactCreateEventComponent.js]: Out of curiosity if you wrap this entire function body in the flag, does size go down?

File: packages/react/src/ReactCreateEventComponent.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    $$typeof: REACT_EVENT_COMPONENT_TYPE,
+    displayName: displayName,
+    props: props,
```
[the remark this answers is not in the document]
Comment (necolas) [reply packages/react/src/ReactCreateEventComponent.js]: should/shouldn't?

[the remark this answers is not in the document]
Comment (trueadm) [reply packages/react/src/ReactCreateEventComponent.js]: Shouldn’t sorry.

File: packages/react/src/ReactCreateEventComponent.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    responder: responder,
+  };
+  if (__DEV__) {
+    Object.freeze(eventComponent);
```
Comment (trueadm) [inline packages/react/src/ReactCreateEventComponent.js]: Given we freeze here. We can remove the freeze operation in FiberComplete phase.

File: packages/shared/ReactTypes.js [hunk 4]
```diff
@@ -136,7 +136,7 @@ export type ReactEventComponentInstance = {|
 
 export type ReactEventComponent = {|
   $$typeof: Symbol | number,
-  displayName?: string,
+  displayName: string,
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/shared/ReactTypes.js]: I.e.

const Press = createEventComponent({ ... });
Press.displayName = “Press”

Just an idea.

**Comment by necolas** [comment]
@trueadm I don't understand why the bundle sizes for individual event components almost doubles with this change

**Comment by trueadm** [comment]
@necolas it’s because it’s bundling all of React with each event module. If you go to bundles.js, you can add React as an external dependency to avoid this.

File: packages/react/src/ReactCreateEventComponent.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import {REACT_EVENT_COMPONENT_TYPE} from 'shared/ReactSymbols';
+
+type EventComponentConfig = {
+  displayName: string,
+  props: null | Object,
+  responder: ReactEventResponder,
+};
+
+export function createEventComponent({
+  displayName,
```
Comment (trueadm) [inline packages/react/src/ReactCreateEventComponent.js]: I think it was maybe better how it was before, as we're creating less objects:

```
export function createEventComponent(responder: ReactEventResponder, displayName?: string)
```

We should also freeze the `responder` too:

```
if (__DEV__) {
  Object.freeze(responder);
}
```

**Comment by necolas** [comment]
Looks like the sizebot comment got deleted. I'll open another PR to get latest report

File: packages/react/src/ReactCreateEventComponent.js (hunk 5 above)
Comment (trueadm) [reply packages/react/src/ReactCreateEventComponent.js]: Sorry, I forgot it got moved: https://github.com/facebook/react/blob/master/packages/react-dom/src/events/DOMEventResponderSystem.js#L466

Comment (trueadm) [reply packages/react/src/ReactCreateEventComponent.js]: > I don't think using an object for a one-off function call is going to be a problem. We used an object for export previously. But happy to change it.

It's not a big deal, but just means less bytes and one less object created which seems better given the API for this function only has 2 arguments now

**Comment by necolas** [comment]
https://github.com/facebook/react/pull/15580

**Pull request closed without merging**